src/Control/Monad/Throws.hs

```diff
@@ -0,0 +1,104 @@
+{-# language FunctionalDependencies #-}
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Control.Monad.Throws where
+
+import Control.Monad.Catch (MonadThrow, MonadCatch, throwM, Exception)
+import Lib
+import GHC.Stack
+import GHC.TypeLits
+import qualified Control.Exception.Safe as Exception
+
+class (Monad m, Exception err) => MonadThrows err m | m -> err where
+    throwAllChecked :: (HasCallStack) => err -> m a
+
+throwChecked :: (MonadThrows err m, ExceptionSubtype err e) => e -> m a
+throwChecked = throwAllChecked . inject
+
+instance
+    ( Exception err, MonadThrow m
+    ) => MonadThrows err (CheckedT err m) where
+    throwAllChecked = throwUnchecked
+
+-- instance
+--     (TypeError ('Text "no"), Exception e, err ~ NoExceptions) =>
+--         MonadThrows e err IO where
+--     throwChecked = throwM
+
+data FooExn = FooExn
+    deriving (Show, Exception)
+
+data BarExn = BarExn
+    deriving (Show, Exception)
+
+blah :: (Throws err (FooExn, BarExn), MonadThrows err m) => Int -> m a
+    -- _ => Int -> _
+blah i
+    | i < 0 =
+        throwChecked FooExn
+    | otherwise =
+        throwChecked BarExn
+
+concretized :: (MonadThrow m, Throws e (FooExn, BarExn)) => CheckedT e m a
+concretized = blah 2
+
+concretized' :: (MonadThrow m) => CheckedT (FooExn || BarExn) m a
+concretized' = blah 2
+
+concretized'' :: (MonadCatch m) => CheckedT BarExn m (Either FooExn a)
+concretized'' = Lib.try @FooExn $ blah 2
+
+-- concretized2 :: (MonadIO m) => m a
+-- concretized2 = liftIO $ blah 2
+
+tryAllChecked ::
+    forall err m a.
+    -- ( Exception e, Exception err,
+    --   MonadCatch m, MonadThrows err m, ExceptionSubtype x err) =>
+    -- (Exception e, MonadCatch m, Exception exn, ExceptionSubtype err exn, MonadThrows err m) =>
+    ( Exception err
+    , MonadCatch m
+    ) =>
+    CheckedT err m a ->
+    m (Either err a)
+tryAllChecked (CheckedT action) = do
+    Exception.try action
+
+tryOneChecked
+    ::  forall e rest m a.
+    ( MonadCatch m
+    , MonadThrows rest m
+    , Exception (e || rest)
+    ) =>
+    CheckedT (e || rest) m a ->
+    m (Either e a)
```

Comment:
This is the holy grail signature I've been looking for. The ability to throw exceptions of type `e` has been plucked. So our input argument can have a type like:

```haskell
blah :: (MonadThrows err m, ExceptionSubtype Foo err, ExceptionSubtype Bar err) => m Int
```

And we can `tryOneChecked @Foo blah` to give back a result with this new type:

```haskell
tryOneChecked @Foo blah
  :: (MonadThrows err m, ExceptionSubtype err Bar)
  => m (Either Foo a)
```

Reply:
OK, so two problems with this signature:

1. `MonadCatch` implies `MonadThrow`, so what we really have is `(MonadThrow m, MonadThrows rest m, MonadCatch m) => ...`. This means that the underlying code implicitly has the ability to do `throwM`, which also implicitly means that calling `try` here would mean that the surrounding code can do `throwM` with impunity. 
2. The `CheckedT` type appears, which means that the `CheckedT` type will have to be given pass-through instances for every class that is used at the given callsites. Not a *huge* issue but not really ideal either.

I posted on the [`exceptions` issue tracker re the first point](https://github.com/ekmett/exceptions/issues/69), quoting myself here:

> The `class (MonadThrow m) => MonadCatch m` implication means that `m` can do `throwM` - and therefore that `MonadThrows rest m` is not a complete accounting for the exceptions that are tracked.
>
> This is a bit philosophical - I suspect a lot of code is going to have `SomeException` as a tracked type, so maybe it's not so bad to have `(MonadThrows err m, ExceptionSubtype err SomeException, ExceptionSubtype err Foo) => ...` given a shorthand of `(MonadThrow m, MonadThrows err m, ExceptionSubtype err Foo) => ...`. 

Maybe a transformer like `newtype ThrowToThrows m a` with `instance (MonadThrows err m, ExceptionSubtype m SomeException) => MonadThrow (ThrowToThrows m)` - that should pluck the relevant constraint out. Still, that's another transformer to write instances for.

Do we need `CheckedT (e || rest) m a` in the signature? Let me see. We need the type variable `e` and `rest` in the input argument so we can try the appropriate exception type. But perhaps a signature like:

```haskell
try
    :: forall e err m a. (MonadThrows err m)
    => m a
    -> m (Either e a)
try action = unsafeRunCheckedT do
  eres <- Checked.tryAll (CheckedT action :: CheckedT (e || err) m a)
  case eres of
    Left (This e) -> pure (Left e)
    Left (That err) -> throwChecked err
    Right a -> pure (Right a)
```

Reply:
the above approach does not seem particularly fruitful (and i need to go on a bike ride)